src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/RazorProjectInfoPublisher.cs

```diff
@@ -212,7 +219,14 @@ internal void Publish(IProjectSnapshot projectSnapshot)
                 // An exception is made for when there's no existing project configuration file because some flashing is preferable to having no TagHelper knowledge.
                 if (ShouldSerialize(projectSnapshot, configurationFilePath))
```

Comment:
It feels odd that the publishing is controlled by this class, because this class is designed to work with serialized files. eg, this `ShouldSerialize` method checks if the file exists, and if it doesn't, it always immediately publishes. With this flag set, the file will never exist, so the logic in this class doesn't apply.

A more problematic example perhaps, in the `DocumentChanged` event, there is a comment that says "document changed is not a good cause for publishing, as it happens while a user types", but the code is checking for the config file existing. Given this will now never exist, it means this class is going to constantly be triggering a publish, despite it being written to do the opposite.

Comments above about the endpoint publisher using the snapshot manager relate to this idea.

Reply:
I might've taken the idea of "just replace file writing with sending a request" too far :) You are right, we should be able to decouple these. Giving it a try.

Reply:
I would have done the same, don't worry. I think starting cautiously was the right approach, to ensure nothing breaks, and really it's only by seeing where that falls down, that we can possibly determine a better shape.

Reply:
Decoupled the classes in https://github.com/dotnet/razor/pull/10182/commits/4ae01c82888ec4acc7403ed3f42c4106bd7f426a and I like it much better, great suggestions, thanks!